keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,345 @@
+
+`password` (optional) - password in *cleartext* if the given private key is protected with it.
+
+For example,
+```bash
+kubectl create secret image-decrypt keysecret --decrypt-secret="/home/user/keys/private.key":"password"
+```
+
+When a user creates a workload, they can choose the desired ImageDecryptSecret to use 
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+  labels:
+    name: nginx
+spec:
+  containers:
+  - name: nginx
+    image: localhost:5000/nginx:enc
+    ports:
+    - containerPort: 80
+  imageDecryptSecrets:
+  - name: keysecret
+```
+
+ImageDecryptSecrets can be added to the `service account` by,
+
+```bash
+kubectl patch serviceaccount <account name> -p '{"imageDecryptSecrets":[{"name":<secret name>]}'
+```
+or while creating a secret account,
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  creationTimestamp: 2015-08-07T22:02:39Z
+  name: default
+  namespace: default
+  selfLink: /api/v1/namespaces/default/serviceaccounts/default
+  uid: 052fb0f4-3d50-11e5-b066-42010af0d7b6
+secrets:
+- name: default-token-uudge
+imagePullSecrets:
+- name: myregistrykey
+imageDecryptSecrets:
+- name: <secret name>
+```
+
+For example,
+```bash
+kubectl patch serviceaccount default -p '{"imageDecryptSecrets":[{"name":"keysecret"}]}'
+```
+
+#### Key Secret Definition
+
+The privake keys are extracted from `ImageDecryptSecret` and passed to the CRI by bundling them in `DecryptParam` as a part of `PullImageRequest` (see, [Image Handler](#Image-handler)) and `CreateContainerRequest` (see, [Container Handler](#Container-handler)):
+
+
+```protobuf
+// DecryptParam represents a single private key (and optional password) that can be sent to the CRI via `PullImageRequest` as well as `CreateContainerRequest`
+message DecryptParam {
+    // key_data represents a private key in format DER/PEM or GPG private keyrings.  These keys will be used by JWE/PKCS7/PGP protocols.
+    bytes key_data = 1;
+    // key_pass represents the (optional) password to unlock the private key
+    bytes key_pass = 2;
+}
+```
+
+API doesn't need to explicitly define the key protocol. Kubernetes passes the given key(s) to runtime _as is_. The runtime should have the logic to handle the key protocol. e.g. The library that will be used by `CRI-O` to handle decryption already has functions to infer the key protocol from the key data, [as seen here](https://github.com/containers/ocicrypt/blob/master/utils/utils.go#L72). 
+
+#### Image Handler
+The list of `DecryptParam` is sent to runtime via `PullImageRequest` to perform image decryption.
+
+```protobuf
+message PullImageRequest {
+    // Spec of the image.
+    ImageSpec image = 1;
+    // Authentication configuration for pulling the image.
+    AuthConfig auth = 2;
+    // Config of the PodSandbox, which is used to pull image in PodSandbox context.
+    PodSandboxConfig sandbox_config = 3;
+    // DecryptParam for the images service of the CRI
+    repeated DecryptParam dcparams = 4;
```

Comment:
Question: is the container runtime supposed to try all the possible decryption keys since multiple could exist?

Reply:
That is my understanding actually, like how image pull secrets work today.

@harche to confirm.

Reply:
If there are multiple possible decryption keys, then the runtime will just iterate over them until it finds the one that can decrypt the symmetric key embedded in OCI annotation of the given image. Also, as Brandon mentioned [earlier](https://github.com/kubernetes/enhancements/pull/1066/files#r328645692), the runtime will use those OCI annotations to figure out the protocol (e.g. jwe, pkcs7 or pgp). 

The way the `ImagePullSecrets` work right now is by passing the [AuthConfig](https://github.com/kubernetes/kubernetes/blob/v1.16.0/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto#L1085) via `PullImageRequest`. `AuthConfig` [message](https://github.com/kubernetes/kubernetes/blob/v1.16.0/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto#L1069) embeds the auth information such as `username`, `password`, `server_address` etc required to pull the image(s) of the pod.

Reply:
We are using standard assymmetric encryption schemes (for wrapping a symmetric key use for layer encryption). The schemes are JWE, PGP, and PKCS11. When a message is encrypted for n recipients using PGP, n encrypted blobs are going to be there and each blob will be associated with a key id used for (asymmetric) encryption. When a private key is used for decryption, the key id is calculated and the decryption attempted only on the blob associated with the same key id, which of course is going to be the only blob leading to a successful decryption (padding is good). So, we defer this problem to the encryption schemes.